File: python/day18.py
from aoc import *


def find_extremes(cubes):
    lower = min(min(c) for c in cubes)
    upper = max(max(c) for c in cubes)
    return range(lower-1, upper+2)


def space_around(cubes):
    limits = find_extremes(cubes)
    seen = cubes.copy()
    queue = [(limits[0],) * 3]
    while queue:
        current = queue.pop()
        for nb in neighbours_3d(*current):
            if nb not in seen and all(coord in limits for coord in nb):
                seen.add(nb)
                queue.append(nb)
    return seen - cubes


def part_1(cubes, neighbours):
    return sum(nb not in cubes for nb in neighbours)


def part_2(cubes, neighbours):
    outside = space_around(cubes)
    return sum(nb in outside for nb in neighbours)


def solve(file=18):
    cubes = set(map(integers, read_input(file)))
    neighbours = [nb for cube in cubes
                     for nb in neighbours_3d(*cube)]
    return part_1(cubes, neighbours), part_2(cubes, neighbours)


print(solve())










def test():
    assert solve("18_test") == (64, 58)
    assert solve() == (3326, 1996)

File: clojure/day18.clj
(ns day18
  (:require aoc
            [clojure.set :as set]))


(defn find-extremes [cubes]
  (let [lower (reduce min (map #(reduce min %) cubes))
        upper (reduce max (map #(reduce max %) cubes))]
    [(dec lower) (inc upper)]))

(defn space-around [cubes]
  (let [[lower upper] (find-extremes cubes)]
    (loop [seen  cubes
           queue (list (repeat 3 lower))]
      (if-let [hd (first queue)]
        (let [valid (->> hd
                         aoc/neighbours-3d
                         (filter (fn [nb]
                                   (and (not (seen nb))
                                        (every? #(<= lower % upper) nb)))))]
          (recur (into seen valid)
                 (into (rest queue) valid)))
        (set/difference seen cubes)))))

(defn parse-input [filename]
  (->> filename
       aoc/read-input
       (map aoc/integers)
       set))

(defn solve
  ([] (solve 18))
  ([input]
   (let [cubes      (parse-input input)
         neighbours (mapcat aoc/neighbours-3d cubes)
         outside    (space-around cubes)]
     [(count (remove cubes neighbours))
      (aoc/count-if outside neighbours)])))


(solve)
